jq255.md

```diff
@@ -0,0 +1,1425 @@
+    U = half((W + J)**2 - Z - T)        # Alternatively: U = J*W
+    E = 2*X - Z - T
+    return (E, Z, U, T)
+```
+
+### Element Equality
+
+The following function determines whether two points `P1` and `P2`
+represent the same group element:
+
+```python
+def same_element(P1, P2):
+    (E1, Z1, U1, T1) = P1
+    (E2, Z2, U2, T2) = P2
+    return U1*E2 == U2*E1
+```
+
+This function handles all cases properly, including when one or both
+of the operands represent the group identity.
+
+A consequence is that a point `P = (E:Z:U:T)` represents the identity
+element in the group if and only if `U = 0`.
+
+### Field to Point Map
+
+We define two field-to-point maps (for jq255e and jq255s, respectively)
+which turn a given field element into a group element (really, a curve
+point) such that the discrete logarithm of the output relatively to the
+conventional generator is unknown. These maps can be implemented in an
+efficient and safe (constant-time) way.
+
+The two jq255e and jq255s curves use distinct implementations, shown
+below. The subsequent pseudocode will assume that the correct
+implementation for the current curve is available under the name
+`map_to_jq255()`.
+
+#### On jq255e
+
+This map uses the constant `sqrtm1 = sqrt(-1)` on the field. Since there are
+two such square roots modulo `q`, we conventionally use the non-negative
+root, i.e.:
+
+```
+sqrtm1 = 7656063742463026568679823572395325799027601838558345258426535816504372595438
+```
+
+The field element `f` is mapped to a point as follows:
+
+```python
+def map_to_jq255e(f):
+    if f == GFq(0):
+        return (GFq(1), GFq(1), GFq(0), GFq(0))   # 0 is mapped to the identity
+    x1 = 4*(f**2) - 7
+    x2 = (4*(f**2) + 7)*sqrtm1   # sqrtm1 is the non-negative square root of -1
+    x0 = 4*f
+    z1 = 64*(f**7) + 176*(f**5) - 308*(f**3) - 343*f
+    z2 = -sqrtm1*(64*(f**7) - 176*(f**5) - 308*(f**3) + 343*f)
+    y0 = 8*(f**2)
+    if is_square(z1):
+        (x, xx, y, yy) = (x1, x0, sqrt(z1), y0)
+    elif is_square(z2):
+        (x, xx, y, yy) = (x2, x0, sqrt(z2), y0)
+    else:
+        (x, xx, y, yy) = (x1*x2, x0**2, sqrt(z1*z2), y0**2)
+    (u, uu) = (x*yy, xx*y)
+    (X, XX) = (-8*(u**2), uu**2)
+    (U, UU) = (2*x*xx*uu, u*(x**2 - 8*(xx**2)))
+    (E, EE) = (X**2 + 2*(XX**2), X**2 - 2*(XX**2))
+    return (E*(UU**2), EE*(UU**2), U*UU*EE, (U**2)*EE)
+```
+
+A constant-time implementation can be achieved by delaying the test
+`f == 0` until the end of the computation, and performing a constant-time
+conditional overwrite of the result with `(GFq(1), GFq(1), GFq(0), GFq(0))`
+in that case. Similarly, `is_square(z1)` and `is_square(z2)` would be
+systematically computed, and constant-time conditional move operations
+performed to get the appropriate operand to the (single) `sqrt()`
+operation, and also set the `x`, `xx`, `y` and `yy` variables
+appropriately.
+
+### On jq255s
+
+On jq255s, we use the Elligator2 map. The field element `f` is mapped to
+a point as follows:
```

Comment:
Why Elligator2 for jq255s and not for jq255e?

Reply:
Elligator2 works on curves that, in short Weierstraß form, have equation `Y^2 = X^3 + A*X + B` with both `A` and `B` being non-zero. For jq255e, the short Weierstraß form leads to `B = 0`, and Elligator2 does not work for that curve, hence the need for something else.

The generic case for curves with `A = 0` or `B = 0` can be covered by finding an isogeny to another curve with `A' != 0` and `B' != 0`, doing the map-to-curve on that other curve, and using the isogeny to get back to our original target curve. This is explained in https://eprint.iacr.org/2019/403 . However, in the case of jq255e, the lowest-degree isogeny is rather complicated, with lots of constants and many multiplications for its computation. The method I use, which I initially found "by hand" when jq255e was called "do255e", leads to simpler code and is also kinda faster, especially on small microcontrollers.

Reply:
Very cool